src/main/java/gregtech/api/capability/impl/CleanroomLogic.java

```diff
@@ -0,0 +1,245 @@
+import gregtech.common.ConfigHolder;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.world.World;
+
+import javax.annotation.Nonnull;
+
+public class CleanroomLogic {
+
+    public static final int BASE_CLEAN_AMOUNT = 5;
+
+    private int maxProgress = 0;
+    private int progressTime = 0;
+
+    private final int minEnergyTier;
+
+    private final MetaTileEntity metaTileEntity;
+    private final boolean hasMaintenance;
+
+    private boolean isActive;
+    private boolean isWorkingEnabled = true;
+    private boolean wasActiveAndNeedsUpdate;
+
+    private boolean hasNotEnoughEnergy;
+
+    public CleanroomLogic(MetaTileEntity metaTileEntity, int minEnergyTier) {
+        this.metaTileEntity = metaTileEntity;
+        this.minEnergyTier = minEnergyTier;
+        this.hasMaintenance = ConfigHolder.machines.enableMaintenance && ((IMaintenance) metaTileEntity).hasMaintenanceMechanics();
+    }
+
+    /**
+     * Performs the actual cleaning
+     * Call this method every tick in update
+     */
+    public void updateLogic() {
+        // cleanrooms which cannot work do nothing
+        if (!this.isWorkingEnabled) return;
+
+        // all maintenance problems not fixed means the machine does not run
+        if (hasMaintenance && ((IMaintenance) metaTileEntity).getNumMaintenanceProblems() > 5) return;
+
+        // drain the energy
+        if (consumeEnergy(true)) {
+            consumeEnergy(false);
+        } else {
+            if (progressTime >= 2) {
+                if (ConfigHolder.machines.recipeProgressLowEnergy) this.progressTime = 1;
+                else this.progressTime = Math.max(1, progressTime - 2);
+
+                hasNotEnoughEnergy = true;
+            }
+            return;
+        }
+
+        if (!this.isActive) setActive(true);
+
+        // increase progress
+        progressTime++;
+        if (progressTime % getMaxProgress() != 0) return;
+        progressTime = 0;
+
+        int amountToClean;
+        if (hasEnoughPower()) amountToClean = BASE_CLEAN_AMOUNT * (getTierDifference() + 1);
+        else amountToClean = -BASE_CLEAN_AMOUNT;
+
+        // each maintenance problem lowers gain by 1 and increases loss by 1
+        if (hasMaintenance) amountToClean -= ((IMaintenance) metaTileEntity).getNumMaintenanceProblems();
+        ((ICleanroomProvider) metaTileEntity).adjustCleanAmount(amountToClean);
+    }
+
+    protected boolean consumeEnergy(boolean simulate) {
+        return ((ICleanroomProvider) metaTileEntity).drainEnergy(simulate);
+    }
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (this.hasNotEnoughEnergy && ((ICleanroomProvider) metaTileEntity).getEnergyInputPerSecond() > 19L * GTValues.VA[((ICleanroomProvider) metaTileEntity).getEnergyTier()]) {
+            this.hasNotEnoughEnergy = false;
+        }
+
+        return true;
```

Comment:
`true if the cleanroom is able to drain energy, else false`
it always returns true